Improvement of time complexity by DFS

mizar opened the pull request:
time complexity: $O(m\times 1.618^n)$ where $m$ is the number of comparators, $n$ is the number of inputs and $\phi=1.618$ is the golden ratio

Verification of networks of about $n=32$ will be easier than before.

see below p.383,
- 3.2 Verification algorithm using the 0-1 principle
- 3.3 Number of computational steps for verification algorithms using the 0-1 principle

refer: Hisayasu Kuroda. (1997). A proposal of Gap Decrease Sorting Network. Trans.IPS.Japan, vol.38, no.3, p.381-389. http://id.nii.ac.jp/1001/00013442/

also, see below p.56, (Here is an example of verifying a network that is not sorted correctly)
- 3.2 Verification algorithm using the 0-1 principle

refer: Hisayasu Kuroda, (1994). A Study of Sorting networks -Gap Decrease Sort-. IPSJ SIG Technical Reports, vol.1994, no.108(1994-HPC-054), p.53-60. http://id.nii.ac.jp/1001/00029885/

mizar committed:
Improvement of time complexity by DFS

mizar commented:
```
python3 sortingnetwork.py --input examples/32-input.cn check
```
```
pypy3 sortingnetwork.py --input examples/32-input.cn check
```

before:
- Python3: (very long hours)
- PyPy3: slightly over 1500 seconds

after:
- Python3: about 18sec
- PyPy3: about 1.5 seconds

```
$ time python3 sortingnetwork.py --input examples/32-input.cn check --show-progress
It is a sorting network!

real    0m18.254s
user    0m18.231s
sys     0m0.003s
$ time pypy3 sortingnetwork.py --input examples/32-input.cn check --show-progress
It is a sorting network!

real    0m1.428s
user    0m1.396s
sys     0m0.030s
```

brianpursley commented:
Thank you for this PR.

It certainly is an interesting approach, and one I was unaware of.

I'm still taking a look at this PR and trying to make sure I understand how it works...

mizar commented:
Thank you for your reply.

It is certainly essential to be able to understand the justification for the algorithm, so I will continue to review the comment text to see if there is a better way to explain it to help with understanding.

brianpursley commented:
I don't think more comments are needed (for my benefit at least), unless you want to explain something further.

I probably should have been more clear, I understand how the code works, but I need to understand why all 2^n binary sequences don't need to be tested in order for the zero-one principle to be applied.  I assumed that all sequences need to be checked in order to be sure that it is a sorting network, but perhaps that isn't true (?).

I saw you linked the research paper, and I will try to read that (using GPT to translate for me).  Is there any other code implementation of this approach that you know of?  I did a quick search the other day but couldn't find any.

I have some other work commitments the next couple of days that will take some time away from me, but I will try to get to it soon.

brianpursley commented:
At first glance, this does seem to work well and it does pass unit tests.  

If you want to you can rebase, since yesterday I updated some of the unit tests to check more thoroughly, but that is up to you.

mizar commented:
- Rebased and squashed.
- Following the 0-1 principle, the result is still trying to verify 2^n different inputs.
- I think the important finding is that the nature of the comparative exchanger and the three-valued logic allows multiple states of 2^n different inputs to be verified all with O(1.618^n) branches.
- So far, I have not found any other published implementations. The Python implementation I tried this time runs on PyPy3 and is only 50% slower than the Rust version I tried, so I think the Python implementation is fast enough.

mizar commented:
The same author's research report presents examples that sort correctly and those that do not. (see below p.56, 3.2 Verification algorithm using the 0-1 principle)

see below p.383,

- 3.2 Verification algorithm using the 0-1 principle
- 3.3 Number of computational steps for verification algorithms using the 0-1 principle

refer: Hisayasu Kuroda. (1997). A proposal of Gap Decrease Sorting Network. Trans.IPS.Japan, vol.38, no.3, p.381-389. http://id.nii.ac.jp/1001/00013442/

also, see below p.56, (Here is an example of verifying a network that is not sorted correctly)

- 3.2 Verification algorithm using the 0-1 principle

refer: Hisayasu Kuroda, (1994). A Study of Sorting networks -Gap Decrease Sort-. IPSJ SIG Technical Reports, vol.1994, no.108(1994-HPC-054), p.53-60. http://id.nii.ac.jp/1001/00029885/

---

In the 0-1 principle, all input examples over `{0,1}` are given, but initially, an input sequence of `{#}` with all indeterminate values is given.

This is a three-valued logic approach.

https://en.wikipedia.org/wiki/Three-valued_logic

The following table shows the transitions for each comparative exchanger.

| (A,B) | (min(A,B),max(A,B)) ||
|:-:|:-:|---|
| (#,#) |   (0,0) or (#,1)    | [branch] A=# and B=#, Split into two cases {alternatively, branch (0,#) or (1,1) works the same way} |
|---|---|  |
| (1,#) |    (#,1) = (B,A)    | [swap] A!=0 and B!=1 and (A!=# or B!=#), Proceed as (B,A) |
| (1,0) |    (0,1) = (B,A)    |  |
| (#,0) |    (0,#) = (B,A)    |  | 
|---|---|  |
| (0,#) |    (0,#) = (A,B)    | [noop] A=0 or B=1, Proceed without changes as (A,B) |
| (#,1) |    (#,1) = (A,B)    |  |
| (0,0) |    (0,0) = (A,B)    |  |
| (0,1) |    (0,1) = (A,B)    |  |
| (1,1) |    (1,1) = (A,B)    |  |

If at the final step or any intermediate stage, the sequence becomes `0…01…1` or `0…0#1…1`, the branch is correctly sorted. This must be confirmed for all branches.

correct example:
```
echo "0:1,2:3,0:2,1:3,1:2" | ./sortingnetwork.py check
```
![image](https://github.com/user-attachments/assets/cb512abb-88aa-428e-8b8f-7faaf604532f)

```
(# # # #)
  (0 0 # #)
    (0 0 0 0) -
    (0 0 # 1) -
  (# 1 # #)
    (# 1 0 0)(0 1 # 0)(0 0 # 1) -
    (# 1 # 1)
      (0 1 0 1)(0 1 0 1)(0 0 1 1) - 
      (# 1 1 1) -
```

The network is found to be correctly sorted, since all branches have transitions of the form `0...01...1` or `0...0#1...1`.

incorrect example:
```
echo "0:1,2:3,1:2,0:1,2:3" | ./sortingnetwork.py check
```
![image](https://github.com/user-attachments/assets/8388a203-0737-41e1-80f7-c4dd286c3926)

```
(# # # #)
  (0 0 # #)
    (0 0 0 0) -
    (0 0 # 1) -
  (# 1 # #)
    (# 1 0 0)(# 0 1 0)(0 # 1 0)(0 # 0 1) x
    (# 1 # 1)
      (# # 1 1)
        (0 0 1 1) -
        (# 1 1 1) -
```

A network example is given where the state of the final stage of a branch may be `0#01`.

This is because neither `0...01...1`, `0...0#1...1`, so it is clear that it cannot be sorted correctly.

---

### time complexity

Branches occur only when the input is `(# #)`. In practice,

```
(# # # # … )
  (0 0 # # … )
    ……
  (# 1 # # … )
    ……
```

The total number of branches $T(n)$ can be written as

$$T(n)=T(n-2)+T(n-1),\quad T(1)=1,\ T(2)=2$$

This is called the Fibonacci sequence,

$$T(n)=\frac{1}{\sqrt{5}}\left(\left(\frac{1+\sqrt{5}}{2}\right)^{n+1}-\left(\frac{1-\sqrt{5}}{2}\right)^{n+1}\right)$$

Here,

$$-1\lt\frac{1-\sqrt{5}}{2}(=-0.6180\cdots)\lt 1$$

So,

$$T(n)\approx\left(\frac{1+\sqrt{5}}{2}\right)^n$$

The value $\displaystyle\frac{1+\sqrt{5}}{2}$ is called the golden ratio, and $T(n)\approx(1.618\cdots)^n$.

Considering the number of comparators $m$, the total number of comparators passed in the 0-1 principle is exactly $m2^n$. In the worst case, when all branches occur at the first $n-1$ comparators.

Thus, the verification time is $O(m\times 1.618^n)$ using the method.

mizar committed:
Fix: improve code comments for clarity

mizar committed:
Fix: correct URL in code comments

brianpursley commented:
Thanks for the detailed explanation.

I think it helped to think about the three-valued logic as it pertains to branching...

So the inputs to a comparator can have 4 possible states:  (0,0) (0,1) (1,0) (1,1)

But after the comparator is applied, the inputs can only be one of 3 possible states:  (0,0), (0,1) (1,1)

In the case of your algorithm, one branch handles (0,0), and the other branch handles *both* (0,1) and (1,1) by using unknown: (#,1)

I agree that when all branches reach `0...01...1` (or `0...0#1...1`), that implies that it is a sorting network.

brianpursley reviewed (approved):
Great job, thanks for the contribution!
